core/src/apps/cardano/sign_tx.py

```diff
@@ -197,10 +206,67 @@ def _validate_outputs(
                 "Each output must have an address field or address_parameters!"
             )
 
+        if output.token_bundle:
+            _validate_token_bundle(output.token_bundle)
+
+        _validate_max_tx_output_size(keychain, output, protocol_magic, network_id)
+
     if total_amount > LOVELACE_MAX_SUPPLY:
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_token_bundle(token_bundle: List[CardanoAssetGroupType]) -> None:
+    seen_policy_ids = set()
+    for token_group in token_bundle:
+        policy_id = bytes(token_group.policy_id)
+
+        if len(policy_id) != MINTING_POLICY_ID_LENGTH:
+            raise INVALID_TOKEN_BUNDLE_OUTPUT
+
+        if policy_id in seen_policy_ids:
+            raise INVALID_TOKEN_BUNDLE_OUTPUT
+        else:
+            seen_policy_ids.add(policy_id)
+
+        if len(token_group.tokens) == 0:
+            raise INVALID_TOKEN_BUNDLE_OUTPUT
+
+        seen_asset_name_bytes = set()
+        for token in token_group.tokens:
+            asset_name_bytes = bytes(token.asset_name_bytes)
+            if len(asset_name_bytes) > MAX_ASSET_NAME_LENGTH:
+                raise INVALID_TOKEN_BUNDLE_OUTPUT
+
+            if asset_name_bytes in seen_asset_name_bytes:
+                raise INVALID_TOKEN_BUNDLE_OUTPUT
+            else:
+                seen_asset_name_bytes.add(asset_name_bytes)
+
+
+def _validate_max_tx_output_size(
+    keychain: seed.Keychain,
+    output: CardanoTxOutputType,
+    protocol_magic: int,
+    network_id: int,
+) -> None:
+    """
+    This limitation is a mitigation measure to prevent sending
+    large (especially change) outputs containing many tokens that Trezor
+    would not be able to spend reliably given that
+    currently the full Cardano transaction is held in-memory.
+    Once Cardano-transaction signing is refactored to be streamed, this
+    limit can be lifted
+    """
+    cborized_output = _cborize_outputs(keychain, [output], protocol_magic, network_id)[
```

Comment:
this is actually thrashing memory by re-generating the CBORized outputs and throwing away the results

is there a reason this isn't inlined in the place where the final transaction is generated?

also you're only checking first output? why?

Reply:
the single output is in the input to this function and the function is invoked in loop for each output, so each output goes through this check indeed. But I agree it can be a bit hard to get at first, so I'll break down the `_cborize_outputs` function implementation into a loop of `_cborize_output` which I can use here directly without such "gymnastics"

Regarding trashing of memory - doesn't garbage collection take care of that? Inlining it into the encoding logic is not trivial as cbor.encode doesn't (nor probably shouldn't) have the context to know whether it's dealing with a tx output right now and decoding the serialized transaction to check how much the output takes doesn't seem reasonable given that by this point already the whole tx is in memory so presumably there's not much space to perform such operation by that point.

Reply:
> doesn't garbage collection take care of that as the memory taken by the cbor representation is referenced only by a local variable in this function?

Well, yes, but actually no. Micropython has mark-and-sweep GC (unlike big Python's refcounting) so you're eating memory only for it to be reclaimed a bit later in a slow sweep run, and possibly fragmenting the available space. We honestly don't have enough RAM for this to be a good solution in any sense, but that's what we have so :woman_shrugging: I believe doing this differently would free up a little more available space for larger transactions, which seems to be something the Cardano app is struggling with.

I wouldn't inline this in `cbor.encode`, but perhaps `_serialize_tx` which already has the cborized structs and can walk the outputs without re-generating them, would be a better choice.

Reply:
To be perfectly clear: let's not micro-optimize, i'll accept the PR even with this code. I am just pointing out a visible trouble spot that should be addressed at some point.

Reply:
I see thanks for the clarification. Then it may make sense to explicitly call garbage collection after validation though of course, that's doesn't guarantee that the memory won't be fragmented, but I agree let's not micro-optimize, especially considering this is supposed to be a temporary workaround until we do the memory-management refactor.